provider/datagen/src/source.rs

```diff
@@ -53,21 +53,25 @@ impl Default for SourceData {
 
 impl SourceData {
     /// The latest CLDR JSON tag that has been verified to work with this version of `icu_datagen`.
+    // We have an unenforced invariant that the data in /tests/data/cldr is generated from this tag.
```

Comment:
Question: what do we gain by moving the data to tests?

Suggestion: If the goal is to remove them from the crates.io package, can we just put them in the ignored files in the Cargo.toml?

Reply:
The way I see it is that this data is testing data for `icu_datagen`, that happens to be used by `make-testdata` as well because you don't want that to talk to the network. If we had this data outside of `icu_datagen` that invariant would be even weaker.

Reply:
I mainly take issue with the notation that this is "testing data for icu_datagen". I see it as "sample data" used for testing all of ICU4X, including icu_datagen. In my mind it's just not semantically correct to put it in `icu_datagen/tests`.

Here are some places where I could see it living:

- `provider/testdata/data/sources`
- `provider/testdata/sources`
- `provider/datagen/sample_data`
- `provider/source_data`

Reply:
> Using datagen/tests/data in make-testdata is a hack to make it work offline

_And_ to make it such that the entire lifecycle of ICU4X data, from source to final output, is easily debuggable.

Reply:
Well it's not that the data behind a tag changes randomly, but yes also that. I don't see how that contradicts any of my points though.

Reply:
I get where you're coming from, and I hope you see where I'm coming from. Can we converge on a mutually agreeable directory path?

Reply:
Your argument is that this data defines everything we do, my argument is that it doesn't because it's equivalent to the tags and merely a performance optimisation. Therefore you want it to have a prominent place, and I want to hide it.

> provider/testdata/data/sources
> provider/testdata/sources

The data is not relevant to testdata, only to its generation's networkiness. Testdata would be the same if we used `SourceData::latest_tested`.

> provider/datagen/sample_data

This data is not used by the datagen crate unless it's built in test mode. Therefore it should be in `tests/`

> provider/source_data

We only have crates at this level of the directory tree, and the only reason this would be in a crate is if we wanted to publish it.

I don't think it should live outside the datagen crate, because having it in there means we can expose it as `SourceData::internal_offline_latest_tested_subset` and point out the invariant that it's similar to `latest_tested`.

Reply:
> This data is not used by the datagen crate unless it's built in test mode

It is `#[doc(hidden)]`, not `#[cfg(test)]`. It seems like a bad style to read _anything_ from the `tests` directory when we're not `#[cfg(test)]`.

I'm fine putting it in the datagen crate. If not `provider/datagen/sample_data`, how about

- `provider/datagen/testdata`
- `provider/datagen/source_data`
- `provider/datagen/internal_source_data`
- `provider/datagen/icu4x_source_data`
- `provider/datagen/testdata_source`
- `provider/datagen/source_data_subset`

Reply:
I'd rather keep it in `tests` and duplicate it in the `testdata-scripts` crate. This is currently the union of what each of those crates need anyway: https://github.com/unicode-org/icu4x/blob/main/tools/testdata-scripts/globs.rs.data#L146

Reply:
That's an interesting proposition, but I don't think it wins us much. We end up with 2 different places in the repo with 2 different subsets of CLDR JSON for no really good reason.

Reply:
`/tools/source-data-cache/{cldr, icuexport}`?

Reply:
How would you access that path from testdata?

Reply:
Relatively

Reply:
Hmm. We also need the data for datagen's own tests, right? And tests need to be runnable on crates.io releases I think. Relative paths don't cut it there.

Reply:
That's why I have it in datagen. I can go back to accessing it through a relative path from testdata-gen, as I agree having something in the `test` directory be exposed through the API is weird.

Reply:
Can we just do datagen/testdata and keep the API function? It is tech debt to have these relative paths.

Reply:
It's not tech debt to refer to crate-local data in a test. I don't want to expose this data on the API.

Reply:
I don't get it. The previous iteration of this PR had the data exported in a `#[doc(hidden)]` API, but loading from the `tests` directory, which we both agreed was a bad design; you said "I agree having something in the test directory be exposed through the API is weird." Your latest iteration changes the API to be `#[cfg(test)]`. I don't see how this is in contradiction with moving the data out of `tests` into its own directory like `testdata` and then going back to the `#[doc(hidden)]` previous status quo.

Reply:
> I agree having something in the test directory be exposed through the API is weird.

My preferred solution to this is to not expose it through the API, not to move it to a publicly available directory.

Reply:
As discussed, `provider/repodata` is the winner.